File: Leetcode_submissions/problems/search_a_2d_matrix/solution.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& arr, int target) {
        if (arr.size() == 0 ||  arr[0].size() == 0 ) return 0;
        int m = arr.size(), n = arr[0].size();
       
        // find row
        int l = 0, r = m-1;
        int row = l + (r-l)/2;
        // printf("row\n");
        while (r>l){
            row = l + (r-l)/2;
            // printf("%d %d %d\n", arr[row][n-1], l , r);
            if (arr[row][n-1] == target) return 1;
            else if (arr[row][n-1] > target){
                r = row;
            }else{
                l = row +1;
            }
        }
        row = l;
        if (arr[row][n-1] == target) return 1;
        // cout << l << " " << row << " " << r << endl;
        
        printf("row %d\n", row);
        
        l = 0, r = n-1;
        while (r>= l){
            int col = l + (r-l)/2;
            // cout << l << " " << col << " "<< r << endl;
            if (arr[row][col] == target) return 1;
            else if (arr[row][col] > target){
                r = col-1;
            }else{
                l = col+1;
            }
        }
        return 0;
    }
};

File: Leetcode_submissions/problems/search_a_2d_matrix/solution.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        
        
        m, n = len(matrix), len(matrix[0])
        l, r = 0, n*m - 1
        
        
        while (r >= l):
            mid = l + int((r-l)/2)
            i, j = int(mid/n), mid % n
            if (matrix[i][j] == target):
                return True
            elif (matrix[i][j] < target):
                l = mid +1
            else:
                r = mid -1
        return False